pkg/controllers/machinesetsync/machineset_sync_controller.go

```diff
@@ -188,3 +322,33 @@ func getInfraMachineTemplateFromProvider(platform configv1.PlatformType) (client
 		return nil, fmt.Errorf("%w: %s", errPlatformNotSupported, platform)
 	}
 }
+
+// setLastTransitionTime determines if the last transition time should be set or updated for a given condition type.
+func setLastTransitionTime(condType machinev1beta1.ConditionType, conditions []machinev1beta1.Condition, conditionAc *machinev1applyconfigs.ConditionApplyConfiguration) {
+	for _, condition := range conditions {
+		if condition.Type == condType {
+			if !hasSameState(&condition, conditionAc) {
+				conditionAc.WithLastTransitionTime(metav1.Now())
+
+				return
+			}
+
+			conditionAc.WithLastTransitionTime(condition.LastTransitionTime)
+
+			return
+		}
+	}
+	// Condition does not exist; set the transition time
+	conditionAc.WithLastTransitionTime(metav1.Now())
+}
+
+// hasSameState returns true if a condition has the same state as a condition
+// apply config; state is defined by the union of following fields: Type,
+// Status, Reason, Severity and Message (it excludes LastTransitionTime).
+func hasSameState(i *machinev1beta1.Condition, j *machinev1applyconfigs.ConditionApplyConfiguration) bool {
+	return i.Type == *j.Type &&
+		i.Status == *j.Status &&
+		i.Reason == *j.Reason &&
+		i.Severity == *j.Severity &&
+		i.Message == *j.Message
```

Comment:
Was this copied from somewhere? Not sure I agree with message being included here, in fact, why is it any more than type and status?

We should look up the definition of the last transition time 🤔 

My concern is that this will constantly tick up for messages that change only in trivial details, like timestamps

Reply:
> Was this copied from somewhere?

Yep, the MAO conditions package. I figured we should keep it consistent with that. https://github.com/openshift/machine-api-operator/blob/master/pkg/util/conditions/gettersetter.go#L160-L168 

I'm not sure if we're including timestamps in messages? Will outputs from `err.Error()` contain timestamps?

Reply:
> We should look up the definition of the last transition time 🤔

For Pods we define it: Timestamp for when the Pod last transitioned from one status to another. 

So, I think it depends on how we define a distinct status, and if we only care about True/False or if messages / reasons matter?

Reply:
>Have you looked at all at how CAPI does this?
Nope, will take a look now

Reply:
Discussing elsewhere: 

In library-go we only use `Status`, so let's keep in line with that. 
https://github.com/openshift/library-go/blame/master/pkg/operator/v1helpers/canonicalize.go#L46


>  My thought here is that you're passing API errors into the condition message, and I'm pretty sure there could be a timestamp somewhere in there, especially if it's an error about updating something that has a timestamp in